src/gha_logs.rs

```diff
@@ -0,0 +1,167 @@
+            self.entries.push_front(entry);
+            Some(&self.entries.front().unwrap().1)
+        } else {
+            None
+        }
+    }
+
+    pub fn put(&mut self, key: String, value: Arc<String>) -> &Arc<String> {
+        let removed = if let Some(pos) = self.entries.iter().position(|(k, _)| k == &key) {
+            self.entries.remove(pos)
+        } else if self.capacity + value.len() as u64 >= MAX_CACHE_CAPACITY_BYTES {
+            self.entries.pop_back()
+        } else {
+            None
+        };
+        if let Some(removed) = removed {
+            self.capacity -= removed.1.len() as u64;
+        }
+        // Insert at the front (most recently used)
+        self.capacity += value.len() as u64;
+        self.entries.push_front((key, value));
+        &self.entries.front().unwrap().1
+    }
+}
+
+pub async fn gha_logs(
+    ctx: Arc<Context>,
+    owner: &str,
+    repo: &str,
+    log_id: &str,
+) -> Result<Response<Body>, hyper::Error> {
+    let res = process_logs(ctx, owner, repo, log_id).await;
+    let res = match res {
+        Ok(r) => r,
+        Err(e) => {
+            tracing::warn!("Unable to serve GHA Logs for {owner}/{repo}#{log_id}: {e:?}");
+            return Ok(Response::builder()
+                .status(StatusCode::INTERNAL_SERVER_ERROR)
+                .body(Body::from(format!("{:?}", e)))
+                .unwrap());
+        }
+    };
+
+    Ok(res)
+}
+
+async fn process_logs(
+    ctx: Arc<Context>,
+    owner: &str,
+    repo: &str,
+    log_id: &str,
+) -> anyhow::Result<Response<Body>> {
+    let log_id = u128::from_str(log_id).context("log_id is not a number")?;
+
+    let repos = ctx
+        .team
+        .repos()
+        .await
+        .context("unable to retrieve team repos")?;
+
+    let Some(repos) = repos.repos.get(owner) else {
+        anyhow::bail!("Organization `{owner}` is not part of team repos")
+    };
+
+    if !repos.iter().any(|r| r.name == repo) {
+        anyhow::bail!("Repository `{repo}` is not part of team repos");
+    }
+
+    let log_uuid = format!("{owner}/{repo}${log_id}");
+
+    let logs = 'logs: {
+        if let Some(logs) = ctx.gha_logs.write().await.get(&log_uuid) {
+            tracing::info!("Cache found for {log_uuid}");
+            break 'logs logs.clone();
+        }
+
+        tracing::info!("Cache NOT found for {log_uuid}");
+        let logs = ctx
+            .github
+            .raw_job_logs(
+                &github::IssueRepository {
+                    organization: owner.to_string(),
+                    repository: repo.to_string(),
+                },
+                log_id,
+            )
+            .await
+            .context("unable to get the raw logs")?;
+
+        ctx.gha_logs
+            .write()
+            .await
+            .put(log_uuid.clone(), logs.into())
+            .clone()
+    };
+
+    let json_logs = serde_json::to_string(&*logs).context("unable to JSON-ify the raw logs")?;
```

Comment:
Micro-opt: we could store the JSON encoded logs in the cache. It should be the ~same size, and it saves one JSON encoding per page load. But that doesn't really matter :)